src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java

```diff
@@ -160,6 +162,83 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
         Assert.assertEquals(pe2.getNextOnGenomeCigarElement(), null);
     }
 
+    @Test(dataProvider = "PrevAndNextTest")
+    public void testImmediateBeforeAndAfterTest(final GATKRead read, final CigarOperator firstOp, final CigarOperator lastOp, final List<CigarOperator> ops) {
+        final AlignmentStateMachine state = new AlignmentStateMachine(read);
+
+        //before the first 'op' from the list
+        state.stepForwardOnGenome();
+        final PileupElement pe1 = state.makePileupElement();
+        Assert.assertEquals(pe1.getAdjacentOperator(PileupElement.Direction.PREV), null);
+        Assert.assertEquals(pe1.getAdjacentOperator(PileupElement.Direction.NEXT), ops.get(0));
+        for (final CigarOperator op : CigarOperator.values()){
+            Assert.assertEquals(pe1.isImmediatelyBefore(op), ops.get(0) == op, op.toString());
+            Assert.assertFalse(pe1.isImmediatelyAfter(op), op.toString());
+        }
+
+        //after the first 'op' from the list
+        state.stepForwardOnGenome();
+        final PileupElement pe2 = state.makePileupElement();
+        Assert.assertEquals(pe2.getAdjacentOperator(PileupElement.Direction.PREV), ops.get(ops.size()-1));
+        Assert.assertEquals(pe2.getAdjacentOperator(PileupElement.Direction.NEXT), null);
+        for (final CigarOperator op : CigarOperator.values()){
+            Assert.assertFalse(pe2.isImmediatelyBefore(op), "immediately before " + op);
+            Assert.assertEquals(pe2.isImmediatelyAfter(op), op ==  ops.get(ops.size()-1), "immediately after " + op);
+        }
```

Comment:
This test is really hard to follow and poorly designed -- can you completely replace it with a new test that:

-constructs PileupElements directly using a read that has a cigar with all possible cigar elements (don't use AlignmentStateMachine, please!), and various offsets into that read
-does direct, straightforward assertions on these PileupElements of the form: `pe.isImmediatelyBefore(CigarOperator.S)`, etc.
-includes test cases in which the current offset is not at the beginning/end of the current cigar element